### ISSUE TITLE
Linting slowing build step by a lot

### ISSUE THREAD

**Issue by liaujianjie** [nusmodifications/nusmods#1239]
### Describe the issue

Currently, linting is part of the webpack build step using `eslint-loader` and it greatly slows the build process.

### To clear the caches for simulating cold build

1. `rm -rf .cache-loader`
2. `rm -rf node_modules/.cache`

### To Profile

Steps for setting up accurate profiling (on development):

1. Install `speed-measure-webpack-plugin` as dev dependency
2. Wrap merged webpack config with the plugin in `webpack.config.dev.js` like so:
```jsx
const SpeedMeasurePlugin = require("speed-measure-webpack-plugin");
const smp = new SpeedMeasurePlugin();

const developmentConfig = smp.wrap(merge([ /* ... */ ]));
```
3. Disable parallelism in webpack config:
```jsx
const developmentConfig = smp.wrap(merge([
  parts.setFreeVariable('process.env.NODE_ENV', 'development'),
  parts.setFreeVariable('process.env.DEBUG_WORKBOX', process.env.DEBUG_WORKBOX),
  commonConfig,
  {
    /* ... */
    parallelism: 1, // <-- here
    /* ... */
  }
]));
```

4. To test with/without linting step, simply comment/uncomment this code block in `webpack.config.common.js`:
```jsx
  // parts.lintJavaScript({
  //   include: parts.PATHS.app,
  // }),
```

### Expected behavior

The build is 45% faster without the `eslint-loader` build step.

### Profiling results

<img width="696" alt="imgonline-com-ua-twotoone-axo45ub9oppfu" src="https://user-images.githubusercontent.com/3143132/47253888-73510680-d48c-11e8-8fae-810b5da73ceb.png">

Left: with linting step
Right: without linting step

### Additional context

Profiled on branch https://github.com/nusmodifications/nusmods/tree/webpack-4.1/www

**Comment by liaujianjie** [comment]
Should probably defer the responsibility of linting to text editor and a pre-commit hook script?

**Comment by ZhangYiJiang** [comment]
Interesting. I tried using the speed measure plugin but it's hard to tell if the plugin itself has a significant impact on performance - the last time I tried the plugin, things took more than twice as long as without. Can you reproduce the difference just measuring it with an external stopwatch? 

The WP4 branch isn't actively being worked on because of the issues mentioned here https://github.com/nusmodifications/nusmods/pull/1052#issuecomment-417209214 

The main reason why we use ESLint loader is to provide lint warnings during development. Maybe we can remove it from the production configs since CI does lints already.

**Comment by liaujianjie** [comment]
I just tested it on the master branch. Build takes about 40% faster without the linting step.

Without linter: `33020ms`, `33734ms`
With linter: `54176ms`, `53188ms`

**Comment by ZhangYiJiang** [comment]
@liaujianjie Do you mean our current Webpack 3 config, removing ESLint loader from the production config?

**Comment by liaujianjie** [comment]
Removing it from the dev config

**Comment by ZhangYiJiang** [comment]
I still think we need eslint-loader for development, since not all editors support ESLint inline. Hot reload performance is the main concern for development, and I think it should still be acceptable right now. 

How about we put it behind a flag that can be disabled by users with editors that have inline ESLint support ? @li-kai @taneliang 

I've removed `lintJavaScript` from prod script in the Webpack 4 branch

**Comment by ZhangYiJiang** [comment]
This is now implemented via #1052. I removed lint and type checks from the production Webpack config - in CI, those are ran in separate stages, and our deployment process is semi-automated, so all code that's pushed into production should already have those checks ran. 

For development, Flow, ESLint and Stylelint are run by default, but can be disabled via env variables `DISABLE_FLOW`, `DISABLE_ESLINT` and `DISABLE_STYLELINT`. We make no assumption about the user's editor, so by default these are enabled in case the user wants to use `nano` to write their code.
